compilation/types/source_maps.go

```diff
@@ -0,0 +1,183 @@
+	ModifierDepth int
+}
+
+// ParseSourceMap takes a source mapping string returned by the compiler and parses it into an array of
+// SourceMapElement objects.
+// Returns the list of SourceMapElement objects.
+func ParseSourceMap(sourceMapStr string) (SourceMap, error) {
+	// Define our variables to store our results in
+	var (
+		sourceMap SourceMap
+		err       error
+	)
+
+	// If our provided source map string is empty, there is no work to be done.
+	if len(sourceMapStr) == 0 {
+		return sourceMap, nil
+	}
+
+	// Separate all the individual source mapping elements
+	elements := strings.Split(sourceMapStr, ";")
+
+	// We use this variable to store "the previous element" because the way
+	// the source mapping works when an element or field is "empty"
+	// the value of the previous element is used.
+	current := SourceMapElement{
+		Index:         -1,
+		Offset:        -1,
+		Length:        -1,
+		FileID:        -1,
+		JumpType:      "",
+		ModifierDepth: 0,
+	}
+
+	// Iterate over all elements split from the source mapping
+	for _, element := range elements {
+		// Set the current index
+		current.Index = len(sourceMap)
+
+		// If the element is empty, we use the previous one
+		if len(element) == 0 {
+			sourceMap = append(sourceMap, current)
+			continue
+		}
+
+		// Split the element fields apart
+		fields := strings.Split(element, ":")
+
+		// If the source range start offset exists, update our current element data.
+		if len(fields) > 0 && fields[0] != "" {
+			current.Offset, err = strconv.Atoi(fields[0])
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		// If the source range length exists, update our current element data.
+		if len(fields) > 1 && fields[1] != "" {
+			current.Length, err = strconv.Atoi(fields[1])
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		// If the source file identifier exists, update our current element data.
+		if len(fields) > 2 && fields[2] != "" {
+			current.FileID, err = strconv.Atoi(fields[2])
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		// If the jump type information exists, update our current element data.
+		if len(fields) > 3 && fields[3] != "" {
+			current.JumpType = SourceMapJumpType(fields[3])
+		}
+
+		// If the modifier call depth exists, update our current element data.
+		if len(fields) > 4 && fields[4] != "" {
+			current.ModifierDepth, err = strconv.Atoi(fields[4])
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		// Append our element to the map
+		sourceMap = append(sourceMap, current)
+	}
+
+	// Return the resulting map
+	return sourceMap, nil
+}
+
+// GetInstructionIndexToOffsetLookup obtains a slice where each index of the slice corresponds to an instruction index,
+// and the element of the slice represents the instruction offset.
+// Returns the slice lookup, or an error if one occurs.
+func (s SourceMap) GetInstructionIndexToOffsetLookup(bytecode []byte) ([]int, error) {
```

Comment:
Doesn't really matter what the choice here is but does this have to be a function of `SourceMap`? It seems like the only thing that is needed from the `SourceMap` is its length. Seems like this can just be a helper function since no aspect of the `SourceMap` is used / changed